plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/editor/autoedit/JSDocEditStrategy.java

```diff
@@ -0,0 +1,142 @@
+			returnString = INDENTATION_STR + RETURN_STR + command.text;
+		}
+		if (returnTypeAndParameterNames.size() > 1) {
+			for (int i = 1; i < returnTypeAndParameterNames.size(); i += 1) {
+				paramString += command.text + INDENTATION_STR + PARAM_STR + returnTypeAndParameterNames.get(i);
+			}
+		}
+		newC.isChange = true;
+		newC.offset = command.offset;
+		newC.text += command.text + INDENTATION_STR;
+		newC.cursorOffset = command.offset + newC.text.length();
+		if (stopTerminal == null && atEndOfLineInput(document, command.offset)) {
+			newC.text += command.text + getRightTerminal();
+		}
+		if (stopTerminal != null && stopTerminal.getOffset() >= command.offset
+				&& util.isSameLine(document, stopTerminal.getOffset(), command.offset)) {
+			String string = document.get(command.offset, stopTerminal.getOffset() - command.offset);
+			if (string.trim().length() > 0)
+				newC.text += string.trim();
+			if (!(returnTypeAndParameterNames.size() == 0)) {
+				newC.text += paramString + command.text + returnString;
+			} else {
+				newC.text += command.text;
+			}
+
+			newC.length += string.length();
+		}
+		return newC;
+	}
+
+	/**
+	 * @returns List<String> otherwise null in case the document is not N4JSDocument or something breaks
+	 */
+	private List<String> getReturnTypeAndParameterNames(IDocument document, IRegion startTerminal) {
+		List<String> retAndfparNames = null;
+		if (document instanceof N4JSDocument) {
+			// index 0: if set to "void", the method has no return value;
+			// fpars start at index 1
+			retAndfparNames = ((N4JSDocument) document).tryReadOnly((xtextResource) -> {
+				Script script = XtextResourceUtils.getScript(xtextResource);
+				ILeafNode node = NodeModelUtils.findLeafNodeAtOffset(NodeModelUtils.findActualNodeFor(script),
+						startTerminal.getOffset());
+				EObject astElement = node.getSemanticElement();
+				List<String> retAndFPars = new ArrayList<>();
+				if (astElement instanceof LiteralOrComputedPropertyName) {
+					if (astElement.eContainer() instanceof N4MethodDeclaration) {
+						N4MethodDeclaration methodDecl = (N4MethodDeclaration) astElement.eContainer();
+						getRetAndFPars(retAndFPars, methodDecl);
+					}
+				} else if (astElement instanceof N4MethodDeclaration) {
+					N4MethodDeclaration methodDecl = (N4MethodDeclaration) astElement;
+					getRetAndFPars(retAndFPars, methodDecl);
+				}
+				return retAndFPars;
+			});
+		}
+		return retAndfparNames;
+	}
+
+	private void getRetAndFPars(List<String> retAndFPars, N4MethodDeclaration methodDecl) {
+		if (methodDecl.getReturnTypeRef() != null) {
+			if ((methodDecl.getReturnTypeRef().getDeclaredType() instanceof VoidTypeImpl)) {
+				retAndFPars.add("void");
+			} else {
+				// e.g. ThisTypeRefNominal or ParameterizedTypeRef
+				retAndFPars.add("return");
+			}
+		} else {
+			// e.g. constructor returnTypeRef is null in methodDecl
+			retAndFPars.add("void");
```

Comment:
Did you test the following case:

```
class C {
    foo() { return 1; }
}
```

In this case, the inferred return type is `any` (which can see in the hover or in the AST-Type model view). 
You can maybe add the following snippet to solve that:

```
else {
			TFunction tfunction = methodDecl.getDefinedFunction();
			if (tfunction != null) {
				TypeRef returnTypeRef = tfunction.getReturnTypeRef();
				if (returnTypeRef != null && !TypeUtils.isVoid(returnTypeRef)) {
					retAndFPars.add("return");
				}
			}
			// e.g. constructor returnTypeRef is null in methodDecl
			retAndFPars.add("void");
		}
```